File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/webrtc/WebRTCClient.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.webrtc

import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import org.webrtc.AudioSource
import org.webrtc.AudioTrack
import org.webrtc.DataChannel
import org.webrtc.DefaultVideoDecoderFactory
import org.webrtc.DefaultVideoEncoderFactory
import org.webrtc.EglBase
import org.webrtc.IceCandidate
import org.webrtc.MediaConstraints
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.RtpReceiver
import org.webrtc.SdpObserver
import org.webrtc.SessionDescription
import org.webrtc.VideoSource
import org.webrtc.VideoTrack

interface WebRTCClientDelegate {
    fun onConnectionStateChanged(state: PeerConnection.IceConnectionState)
    fun onIceCandidateGenerated(candidate: IceCandidate)
    fun onRemoteVideoTrackReceived(track: VideoTrack)
    fun onRemoteVideoTrackRemoved(track: VideoTrack)
}

class WebRTCClient(private val context: Context) {
    companion object {
        private const val TAG = "WebRTCClient"
    }

    var delegate: WebRTCClientDelegate? = null

    private var peerConnectionFactory: PeerConnectionFactory? = null
    private var peerConnection: PeerConnection? = null
    private var localVideoTrack: VideoTrack? = null
    private var localAudioTrack: AudioTrack? = null
    private var remoteVideoTrack: VideoTrack? = null
    private var videoSource: VideoSource? = null
    private var customCapturer: CustomVideoCapturer? = null

    val eglBase: EglBase = EglBase.create()

    fun setup(iceServers: List<PeerConnection.IceServer>?) {
        // Initialize PeerConnectionFactory
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(context)
                .setEnableInternalTracer(false)
                .createInitializationOptions()
        )

        val encoderFactory = DefaultVideoEncoderFactory(eglBase.eglBaseContext, true, true)
        val decoderFactory = DefaultVideoDecoderFactory(eglBase.eglBaseContext)

        peerConnectionFactory = PeerConnectionFactory.builder()
            .setVideoEncoderFactory(encoderFactory)
            .setVideoDecoderFactory(decoderFactory)
            .createPeerConnectionFactory()

        // Configure ICE servers
        val rtcConfig = PeerConnection.RTCConfiguration(
            iceServers ?: listOf(
                PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer()
            )
        )
        rtcConfig.sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
        rtcConfig.continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY

        // Create peer connection
        peerConnection = peerConnectionFactory?.createPeerConnection(
            rtcConfig,
            object : PeerConnection.Observer {
                override fun onSignalingChange(state: PeerConnection.SignalingState?) {
                    Log.d(TAG, "Signaling state: $state")
                }

                override fun onIceConnectionChange(state: PeerConnection.IceConnectionState?) {
                    Log.d(TAG, "ICE connection state: $state")
                    state?.let { delegate?.onConnectionStateChanged(it) }
                }

                override fun onIceConnectionReceivingChange(receiving: Boolean) {}

                override fun onIceGatheringChange(state: PeerConnection.IceGatheringState?) {
                    Log.d(TAG, "ICE gathering state: $state")
                }

                override fun onIceCandidate(candidate: IceCandidate?) {
                    candidate?.let {
                        // Log candidate type for debugging
                        when {
                            it.sdp.contains("relay") -> Log.d(TAG, "ICE candidate: RELAY (TURN)")
                            it.sdp.contains("srflx") -> Log.d(TAG, "ICE candidate: SERVER REFLEXIVE (STUN)")
                            it.sdp.contains("host") -> Log.d(TAG, "ICE candidate: HOST (local)")
                        }
                        delegate?.onIceCandidateGenerated(it)
                    }
                }

                override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>?) {}

                override fun onAddStream(stream: MediaStream?) {
                    Log.d(TAG, "Remote stream added: ${stream?.videoTracks?.size} video, ${stream?.audioTracks?.size} audio")
                    stream?.videoTracks?.firstOrNull()?.let { track ->
                        remoteVideoTrack = track
                        delegate?.onRemoteVideoTrackReceived(track)
                    }
                }

                override fun onRemoveStream(stream: MediaStream?) {
                    Log.d(TAG, "Remote stream removed")
                    remoteVideoTrack?.let { track ->
                        remoteVideoTrack = null
                        delegate?.onRemoteVideoTrackRemoved(track)
                    }
                }

                override fun onDataChannel(channel: DataChannel?) {}
                override fun onRenegotiationNeeded() {
                    Log.d(TAG, "Negotiation needed")
                }
                override fun onAddTrack(receiver: RtpReceiver?, streams: Array<out MediaStream>?) {}
            }
        )

        // Create video source + track with custom capturer
        val factory = peerConnectionFactory ?: return
        videoSource = factory.createVideoSource(false) // false = not a screen cast
        customCapturer = CustomVideoCapturer().apply { initialize(videoSource!!) }
        localVideoTrack = factory.createVideoTrack("video0", videoSource).apply {
            setEnabled(true)
        }
        peerConnection?.addTrack(localVideoTrack, listOf("stream0"))

        // Create audio track
        val audioConstraints = MediaConstraints()
        val audioSource: AudioSource = factory.createAudioSource(audioConstraints)
        localAudioTrack = factory.createAudioTrack("audio0", audioSource).apply {
            setEnabled(true)
        }
        peerConnection?.addTrack(localAudioTrack, listOf("stream0"))
    }

    fun pushVideoFrame(bitmap: Bitmap) {
        customCapturer?.pushFrame(bitmap)
    }

    fun createOffer(onSuccess: (SessionDescription) -> Unit) {
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "true"))
        }

        peerConnection?.createOffer(object : SdpObserver {
            override fun onCreateSuccess(sdp: SessionDescription?) {
                sdp?.let { offer ->
                    peerConnection?.setLocalDescription(object : SdpObserver {
                        override fun onSetSuccess() {
                            onSuccess(offer)
                        }
                        override fun onSetFailure(error: String?) {
                            Log.e(TAG, "Failed to set local description: $error")
                        }
                        override fun onCreateSuccess(p0: SessionDescription?) {}
                        override fun onCreateFailure(p0: String?) {}
                    }, offer)
                }
            }
            override fun onCreateFailure(error: String?) {
                Log.e(TAG, "Failed to create offer: $error")
            }
            override fun onSetSuccess() {}
            override fun onSetFailure(p0: String?) {}
        }, constraints)
    }

    fun setRemoteSdp(sdp: SessionDescription, onComplete: (String?) -> Unit) {
        peerConnection?.setRemoteDescription(object : SdpObserver {
            override fun onSetSuccess() {
                onComplete(null)
            }
            override fun onSetFailure(error: String?) {
                onComplete(error)
            }
            override fun onCreateSuccess(p0: SessionDescription?) {}
            override fun onCreateFailure(p0: String?) {}
        }, sdp)
    }

    fun addRemoteCandidate(candidate: IceCandidate, onComplete: (String?) -> Unit) {
        val success = peerConnection?.addIceCandidate(candidate) ?: false
        onComplete(if (success) null else "Failed to add ICE candidate")
    }

    fun muteAudio(mute: Boolean) {
        localAudioTrack?.setEnabled(!mute)
    }

    fun close() {
        localVideoTrack?.setEnabled(false)
        localAudioTrack?.setEnabled(false)
        remoteVideoTrack = null
        customCapturer?.dispose()
        customCapturer = null
        peerConnection?.close()
        peerConnection = null
        peerConnectionFactory?.dispose()
        peerConnectionFactory = null
        Log.d(TAG, "Peer connection closed")
    }
}

File: samples/CameraAccess/CameraAccess/WebRTC/WebRTCClient.swift
import Foundation
import WebRTC

protocol WebRTCClientDelegate: AnyObject {
  func webRTCClient(_ client: WebRTCClient, didChangeConnectionState state: RTCIceConnectionState)
  func webRTCClient(_ client: WebRTCClient, didGenerateCandidate candidate: RTCIceCandidate)
  func webRTCClient(_ client: WebRTCClient, didReceiveRemoteVideoTrack track: RTCVideoTrack)
  func webRTCClient(_ client: WebRTCClient, didRemoveRemoteVideoTrack track: RTCVideoTrack)
}

/// Manages RTCPeerConnection, video/audio tracks, and SDP negotiation.
/// Video uses a custom capturer (fed by DAT SDK frames). Audio uses WebRTC's native engine.
class WebRTCClient: NSObject {
  weak var delegate: WebRTCClientDelegate?

  private let factory: RTCPeerConnectionFactory
  private var peerConnection: RTCPeerConnection?
  private var videoSource: RTCVideoSource!
  private var videoCapturer: CustomVideoCapturer!
  private var localVideoTrack: RTCVideoTrack?
  private var localAudioTrack: RTCAudioTrack?
  private(set) var remoteVideoTrack: RTCVideoTrack?

  override init() {
    RTCInitializeSSL()
    let encoderFactory = RTCDefaultVideoEncoderFactory()
    let decoderFactory = RTCDefaultVideoDecoderFactory()
    self.factory = RTCPeerConnectionFactory(
      encoderFactory: encoderFactory,
      decoderFactory: decoderFactory
    )
    super.init()
  }

  func setup(iceServers: [RTCIceServer]? = nil) {
    let config = RTCConfiguration()
    config.iceServers = iceServers ?? [RTCIceServer(urlStrings: WebRTCConfig.stunServers)]
    config.sdpSemantics = .unifiedPlan
    config.continualGatheringPolicy = .gatherContinually

    let constraints = RTCMediaConstraints(
      mandatoryConstraints: nil,
      optionalConstraints: ["DtlsSrtpKeyAgreement": "true"]
    )

    peerConnection = factory.peerConnection(
      with: config, constraints: constraints, delegate: self
    )

    createMediaTracks()
  }

  private func createMediaTracks() {
    // Video track — custom source fed by DAT SDK frames
    videoSource = factory.videoSource()
    videoCapturer = CustomVideoCapturer(delegate: videoSource)
    localVideoTrack = factory.videoTrack(with: videoSource, trackId: "video0")
    localVideoTrack?.isEnabled = true
    peerConnection?.add(localVideoTrack!, streamIds: ["stream0"])

    // Audio track — WebRTC native audio (handles mic capture, AEC, playback)
    let audioConstraints = RTCMediaConstraints(
      mandatoryConstraints: nil, optionalConstraints: nil
    )
    let audioSource = factory.audioSource(with: audioConstraints)
    localAudioTrack = factory.audioTrack(with: audioSource, trackId: "audio0")
    localAudioTrack?.isEnabled = true
    peerConnection?.add(localAudioTrack!, streamIds: ["stream0"])
  }

  /// Called by ViewModel to push video frames from DAT SDK / iPhone camera.
  func pushVideoFrame(_ image: UIImage) {
    videoCapturer?.pushFrame(image)
  }

  // MARK: - SDP Negotiation

  func createOffer(completion: @escaping (RTCSessionDescription) -> Void) {
    let constraints = RTCMediaConstraints(
      mandatoryConstraints: [
        "OfferToReceiveAudio": "true",
        "OfferToReceiveVideo": "true",
      ],
      optionalConstraints: nil
    )
    peerConnection?.offer(for: constraints) { [weak self] sdp, error in
      guard let sdp else {
        NSLog("[WebRTC] Failed to create offer: %@", error?.localizedDescription ?? "unknown")
        return
      }
      self?.peerConnection?.setLocalDescription(sdp) { error in
        if let error {
          NSLog(
            "[WebRTC] Failed to set local description: %@", error.localizedDescription)
        } else {
          completion(sdp)
        }
      }
    }
  }

  func set(remoteSdp: RTCSessionDescription, completion: @escaping (Error?) -> Void) {
    peerConnection?.setRemoteDescription(remoteSdp, completionHandler: completion)
  }

  func set(remoteCandidate: RTCIceCandidate, completion: @escaping (Error?) -> Void) {
    peerConnection?.add(remoteCandidate, completionHandler: completion)
  }

  func muteAudio(_ mute: Bool) {
    localAudioTrack?.isEnabled = !mute
  }

  func close() {
    localVideoTrack?.isEnabled = false
    localAudioTrack?.isEnabled = false
    remoteVideoTrack = nil
    peerConnection?.close()
    peerConnection = nil
    NSLog("[WebRTC] Peer connection closed")
  }

  deinit {
    RTCCleanupSSL()
  }
}

// MARK: - RTCPeerConnectionDelegate

extension WebRTCClient: RTCPeerConnectionDelegate {
  func peerConnection(
    _ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState
  ) {
    NSLog("[WebRTC] Signaling state: %d", stateChanged.rawValue)
  }

  func peerConnection(
    _ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState
  ) {
    NSLog("[WebRTC] ICE connection state: %d", newState.rawValue)
    delegate?.webRTCClient(self, didChangeConnectionState: newState)
  }

  func peerConnection(
    _ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState
  ) {
    NSLog("[WebRTC] ICE gathering state: %d", newState.rawValue)
  }

  func peerConnection(
    _ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate
  ) {
    // Log candidate type for debugging NAT traversal
    let sdp = candidate.sdp
    if sdp.contains("relay") {
      NSLog("[WebRTC] ICE candidate: RELAY (TURN)")
    } else if sdp.contains("srflx") {
      NSLog("[WebRTC] ICE candidate: SERVER REFLEXIVE (STUN)")
    } else if sdp.contains("host") {
      NSLog("[WebRTC] ICE candidate: HOST (local)")
    }
    delegate?.webRTCClient(self, didGenerateCandidate: candidate)
  }

  func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
    NSLog("[WebRTC] Remote stream added with %d audio tracks, %d video tracks",
          stream.audioTracks.count, stream.videoTracks.count)
    if let videoTrack = stream.videoTracks.first {
      remoteVideoTrack = videoTrack
      delegate?.webRTCClient(self, didReceiveRemoteVideoTrack: videoTrack)
    }
  }

  func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {
    NSLog("[WebRTC] Remote stream removed")
    if let track = remoteVideoTrack {
      remoteVideoTrack = nil
      delegate?.webRTCClient(self, didRemoveRemoteVideoTrack: track)
    }
  }

  func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {
    NSLog("[WebRTC] Negotiation needed")
  }

  func peerConnection(
    _ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]
  ) {}

  func peerConnection(
    _ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel
  ) {}
}
